### PR TITLE
Add ability to trigger manual builds

### LINKED ISSUE #83: Add support to trigger pipelines manually
# Todo
- add api endpoint
- add button to ui
- add cli command

### LINKED ISSUE #240: Create build from API?
It looks like the Drone API has an endpoint where you POST to `/repos/<>/<>/builds` and the thing does the whoozit.

This would be very helpful, but I'm guessing it was implemented sometime after the fork. If I'm right, rather than just overlooking something, it might be worth implementing at some other endpoint. Personally, I see where putting it at "POST to /builds" saved time and trouble, but I think it's a little counterintuitive 🤷‍♂️

### PR DESCRIPTION
**Pull request by ennetech**
closes #83 
closes #240 

Changelog:
- [x] UI: button to trigger manual build
- [x] SERVER: ability to inject variables for a specific build
- [x] SERVER: trigger action for build
- [x] Updated gitea dep to allow darwin builds
- [x] UI: style the form & localization
- [x] SERVER: auto extract of last commit id of branch
- [x] SERVER: debug why adding a new event "manual" leads to only clone step being run
- [x] Error handling for api endpoint
- [x] MIssing fields for build
- [x] Persist variables on build restart
- [x] cli command
- [x] find a home for ManualBuildReq
- [x] security checks
- [x] hide button if not available to user

### CODE REVIEW

File: server/api/build.go
Comment (ennetech) [reply server/api/build.go:57]: done

File: server/api/build.go
Comment (ennetech) [reply server/api/build.go]: done

File: server/api/build.go
Comment (ennetech) [reply server/api/build.go]: done

**Comment by ennetech** [comment]
ok major things have been sorted out, i'll start working on the ui

File: server/model/build.go
Comment (6543) [inline server/model/build.go:54]: Why do we need varibles in build?

Comment (ennetech) [reply server/model/build.go:54]: some use cases:
- when you have one repo that generates multiple images and you only want to trigger one
- disabling some feature flags without having to commit

File: web/src/views/repo/RepoBuild.vue
Comment (anbraten) [inline web/src/views/repo/RepoBuild.vue:85]: Could you migrate this to composition api with setup? If you need help just ask, I would be happy to support you. 

For starting a manual pipeline I would suggest having a button like this at the top:

image

Which opens a popup like this:

Screenshot from 2022-09-02 19-14-28

File: web/src/views/repo/RepoBuild.vue
Comment (anbraten) [inline web/src/views/repo/RepoBuild.vue:66]:
```suggestion
import SelectField from '~/components/form/SelectField.vue';
import Panel from '~/components/layout/Panel.vue';
```

File: web/src/views/repo/RepoBuild.vue
Comment (anbraten) [inline web/src/views/repo/RepoBuild.vue:86]: Please convert this to `async await` syntax.

**Comment by ennetech** [comment]
Did a general restyle:
image
and fixed all type errors // removed promise etc

only minor things remain to be 100% composition api

File: web/src/views/repo/RepoBuild.vue
Comment (ennetech) [reply web/src/views/repo/RepoBuild.vue]: i did not find any popup to gather inspiration from inside woodpecker codebase, do you have any suggestion?

I can still implement that from scratch, but i would like to avoid reinventing the wheel

File: server/api/build.go
Comment (ennetech) [reply server/api/build.go]: done

File: web/src/views/repo/RepoBuild.vue
Comment (ennetech) [reply web/src/views/repo/RepoBuild.vue]: @anbraten migrated all to composition api

Comment (anbraten) [reply web/src/views/repo/RepoBuild.vue]: I don't think we have a popup yet, but it should be quite simple to implement base on something like this:

```vue
<template>
  <!-- overlay -->
  <div
    v-if="open"
    class="fixed bg-gray-900 opacity-80 left-0 top-0 right-0 bottom-0 z-500 print:hidden"
    @click="$emit('close')"
  />
  <!-- overlay end -->
  <transition :name="`slide-${position}`" class="print:hidden fixed flex left-0 top-0 right-0 bottom-0">
    <div
      v-if="open"
      class="m-auto flex flex-col bg-gray-200 shadow-all z-1000 max-w-64"
    >
      <slot />
    </div>
  </transition>
</template>

<script lang="ts">
import { defineComponent, PropType, ref, toRef, watch } from 'vue';

export default defineComponent({
  name: 'Popup',

  props: {
    open: {
      type: Boolean,
    },
  },

  emits: {
    close: () => true,
  },
});
</script>

<style scoped>
.slide-right-enter-active,
.slide-right-leave-active {
  transition: all 0.3s ease;
}
.slide-right-enter-from,
.slide-right-leave-to {
  transform: translate(100%, 0);
}

.slide-bottom-enter-active,
.slide-bottom-leave-active {
  transition: all 0.3s ease;
}
.slide-bottom-enter-from,
.slide-bottom-leave-to {
  transform: translate(0, 100%);
}
</style>
```

Comment (ennetech) [reply web/src/views/repo/RepoBuild.vue]: for the moment i have added the button to the right that loads the page in the router-view as before:
image

a popup is not so usefull as after trigger the page will be redirected to the created pipeline

File: server/api/build.go
Comment (ennetech) [inline server/api/build.go:44]: what should be the correct place for this struct?

**Comment by ennetech** [comment]
added function in go client and cli command, i would need a suggestion about where to place ManualBuildReq, in the meantime i'll refactor and add the names in the ui

**Comment by 6543** [comment]
I thin this pull only will solve #240 but not #83 as that is about manually imitating other events if i understand correctly

**Comment by 6543** [comment]
& since it was added to the default event filter, docs have to be adjusted

**Comment by ennetech** [comment]
@6543 actually it passes a Build to pipeline.Create, so it doesn't fake any existing event, i created the "manual" event

The page to be updated is https://woodpecker-ci.org/docs/usage/conditional-execution ?
Also, where the manual trigger should be documented in a new page or in a existing one?

**Comment by ennetech** [comment]
> I believe we can extend this later to close #83 to support event injection.

do you mean like "deploy", "rollback" and similar?

**Comment by ennetech** [comment]
> This is completely missing any security checks.
> 
> 1. The UI option should be hidden if you can't use it.
> 2. If you're a guest (not logged in), a nil pointer exception occurs in `server/api/build.go:67`
> 3. If you're logged in as another user that does not have write access to the repo, the user still can use "Run build".

1. What is the current way of delivering configuration to the frontend?
2. Should be the building be available to guests?
3. What's the correct ACL for this? Only write on repo?

**Comment by 6543** [comment]
You dont have to rebase&force-push just merge master into, pulls get squash-merged

**Review by 6543 — DISMISSED** [review]
As per comment

File: woodpecker-go/woodpecker/client.go
Comment (6543) [inline woodpecker-go/woodpecker/client.go:219]: Type should be declared in types.go

File: server/pipeline/items.go
Comment (6543) [inline server/pipeline/items.go:63]: I personaly still dont like it to add a seperate feature in one pull

File: server/model/build.go
Comment (6543) [inline server/model/build.go:54]: If you trigger it from the outside why would you store that dater in database?

File: woodpecker-go/woodpecker/client.go
Comment (6543) [reply woodpecker-go/woodpecker/client.go:219]: I also would not call it manual build option, as if we gona add otjer events support it is just "create a build", so this func should get a option struct

**Comment by 6543** [comment]
Hope I doesnt sound to harsh, i just like to make sure it hasnt to be refactored later ...

**Comment by ennetech** [comment]
> Hope I doesnt sound to harsh, i just like to make sure it hasnt to be refactored later ...

No worries, better now than later

**Comment by ennetech** [comment]
> > 1. What is the current way of delivering configuration to the frontend?
> 
> There's an API endpoint that's queried once. You can look at the "cancel"/"restart" buttons, they should use it as well.
> 
> > 2. Should be the building be available to guests?
> 
> No.
> 
> > 3. What's the correct ACL for this? Only write on repo?
> 
> Only allow it if you have write permissions on the repo.

regarding 2 & 3, should this be solved adding session.MustPush in the api.go?

**Comment by anbraten** [comment]
> > > 1. What is the current way of delivering configuration to the frontend?
> > 
> > There's an API endpoint that's queried once. You can look at the "cancel"/"restart" buttons, they should use it as well.
> > 
> > > 2. Should be the building be available to guests?
> > 
> > No.
> > 
> > > 3. What's the correct ACL for this? Only write on repo?
> > 
> > Only allow it if you have write permissions on the repo.
> 
> regarding 2 & 3, should this be solved adding session.MustPush in the api.go?

yes

**Comment by ennetech** [comment]
perfect, all pushed, only things need to sort out is:
- the hiding of the button on the ui
- decide if a separate pull request for variables is needed

File: server/api/build.go
Comment (lafriks) [inline server/api/build.go:24]: Invalid import position should be together with line 30

File: server/model/build.go
Comment (6543) [inline server/model/build.go:54]: I would like to rename it to "CustomBuildVars or similar ... as it does not stor all vars just user defined single run custom vars

File: server/pipeline/items.go
Comment (6543) [inline server/pipeline/items.go:64]: This can overwrite all existing one, is this intemded behaviour?

File: web/src/assets/locales/en.json
Comment (6543) [inline web/src/assets/locales/en.json:32]:
```suggestion
            "title": "Additional Build variables",
```

File: web/src/assets/locales/en.json
Comment (6543) [inline web/src/assets/locales/en.json:33]:
```suggestion
            "desc": "Specify additional variables that can overwrite XYZ... env vars",
```
Need better desc.

File: woodpecker-go/woodpecker/interface.go
Comment (6543) [inline woodpecker-go/woodpecker/interface.go:74]: Should use buildOptions

File: server/model/build.go
Comment (anbraten) [reply server/model/build.go:54]: Maybe `additional variables` as you suggested for the ui. 

We could also use those later for other features as for example the "comment" events where one auto added variable could be `comment_event_type: created`.

File: server/pipeline/items.go
Comment (anbraten) [reply server/pipeline/items.go:64]: I would say it should not be able to override existing vars as that could be a security risk as well or we should think about giving them a prefix.

File: web/src/router.ts
Comment (anbraten) [inline web/src/router.ts:58]:
```suggestion
        path: 'pipeline/manual',
        name: 'repo-manual-pipeline',
```

File: web/src/views/repo/RepoBuild.vue
Comment (anbraten) [inline web/src/views/repo/RepoBuild.vue:60]: Do we need this? You can already use `CI_BUILD_EVENT=manual` for it.

File: web/src/views/repo/RepoBuild.vue
Comment (anbraten) [inline web/src/views/repo/RepoBuild.vue:96]:
```suggestion
```

Still the same so not needed.

File: server/api/build.go
Comment (anbraten) [inline server/api/build.go:64]: What about using the last commit message for this again and maybe adding this prefix?

**Comment by anbraten** [comment]
I pushed some adjustments to the UI and did some cleanup (Hopefully fine for you). 

Some notes:
- We should keep on naming it `pipeline` instead of `build` in all user facing parts (migrating the technical code has to be done at once at some later point 🙈)
- We should rename `manual` to sth like `manual_pipeline` in most places of the code like icon name, translation, component name (except the event itself)
- We need to add the `manual` event type to the secrets options
- I am still not super happy with the panel being a page / tab instead of a popup etc

image

File: server/pipeline/items.go
Comment (6543) [reply server/pipeline/items.go:64]: How can you abuse it?

File: web/src/views/repo/RepoBuild.vue
Comment (anbraten) [reply web/src/views/repo/RepoBuild.vue:20]:
```suggestion
              <Button type="submit" text="X" class="ml-auto" @click="deleteVar(name)">
                <i-la-times />
              </Button>
```

**Comment by ennetech** [comment]
I've tried to implement the popup using @anbraten suggestion (https://github.com/woodpecker-ci/woodpecker/pull/1156#discussion_r961915892) but a little bit of frontend magic is needed to:
- proper position the overlay (sometimes the click outside does not work)
- fix the responsive layout
- general refactor of the file position (at the moment i have just imported the "old" view)

File: server/pipeline/items.go
Comment (anbraten) [reply server/pipeline/items.go:64]: You could bypass some filter based on env vars, but I noticed that as you need push access to trigger a manual pipeline this is not a problem.

File: server/api/build.go
Comment (anbraten) [reply server/api/build.go:30]: Its currently using the type `BuildOptions` from it.

**Comment by anbraten** [comment]
@ennetech Thanks for the changes. I can have a look at it again and try to improve the UI in your mentioned points a bit if you want me to.

**Comment by ennetech** [comment]
> @ennetech Thanks for the changes. I can have a look at it again and try to improve the UI in your mentioned points a bit if you want me to.

Yes, thank you!

**Comment by ennetech** [comment]
Awesome @anbraten, thanks!

File: server/api/build.go
Comment (6543) [reply server/api/build.go:30]: well I'm aware of that :laughing:  - we should avoid this dependency graph 

I'll already try to add some dependency graph linting rules but for now it dont work  100% ...

File: server/api/build.go
Comment (6543) [inline server/api/build.go:57]: can we move create Build out of  CreateBuildOptions into it's own encapsulated func so we can unit test it?

File: cli/build/build_create.go
Comment (6543) [inline cli/build/build_create.go:52]: SplitN(vaz,"=",1)

Comment (6543) [reply cli/build/build_create.go:52]: https://play.golang.com/p/atzxoNHjVm1

File: cli/build/build.go
Comment (6543) [reply cli/build/build.go:26]: just exec  `make docs` and commit the result

**Comment by ennetech** [comment]
@6543 @anbraten what is left to do?

File: server/api/build.go
Comment (anbraten) [inline server/api/build.go:75]: Would still like to change this to the last commit message of the corresponding branch, but we could refactor this in separate PR together with the remote apis.

**Review by anbraten — APPROVED** [review]

**Comment by ennetech** [comment]
> LGTM if you address the last requested change (https://github.com/woodpecker-ci/woodpecker/pull/1156/files#r966210174)

Where the struct should be duplicated?

**Comment by ennetech** [comment]
@anbraten @qwerty287 @6543 i've moved the dependency away, but i don't think is the correct place

**Comment by 6543** [comment]
once merged we should create issues that document the followups

**Comment by 6543** [comment]
> once merged we should create issues that document the followups

->  #1216

**Comment by 6543** [comment]
are there any other?

**Comment by 6543** [comment]
->  #1222

**Pull request merged**
